sim-administration/simmanager/src/main/kotlin/org/ostelco/simcards/admin/PeriodicProvisioningTask.kt

```diff
@@ -36,6 +35,9 @@ import kotlin.math.min
  * there will be a number of profiles available for quick allocation to customers
  * without having to synchronously wait for a profile to be provisioned by these
  * two.
+ *
+ * TODO:  This is so incredibly complicated, and there is no need for it to be.
+ *        Just fix it!
```

Comment:
As part of this. Noted that the
```
            if (simEntry.id == null) { // TODO: This idiom is _bad_, find something better!
                AdapterError("simEntry.id == null for simEntry = '$simEntry'.").left()
            } else
```
(line 84-86) was already done by the caller, and that these lines therefore can be removed (unless I have overlooked something...).

Also on line 150 `return "ok".right()`. Would be better to return `Unit.right()` I think.

Reply:
Generally I think that readability could be improved if some of the code inside `for` loops was moved out as functions?

Reply:
I will leave this for now, since it is my intention to refactor this code extensively later, the for loop is only one of many things that will be fixed.  The "Ok" text is another.  The method should return a result that shall be returned to the caller so that the caller can see a summary of  what the task actually accomplished.

Reply:
A rule of thumb is that the level of parenthesis shouldn't be more than three or so, so clearly this method, and several others in this file, break this rule and clear targets for refactoring.

Reply:
I have made TODOs  of these issues in my working copy and will address them in future PRs, possibly even the next one.